diff --git a/client.c b/client.c
index 3d6d82b..b653047 100644
--- a/client.c
+++ b/client.c
@@ -47,12 +47,13 @@ static void client_setup_terminal(void) {
 	cur_term.c_cc[VMIN] = 1;
 	cur_term.c_cc[VTIME] = 0;
 	tcsetattr(STDIN_FILENO, TCSANOW, &cur_term);
-
+#if 0
 	if (!alternate_buffer) {
 		printf("\033[?1049h\033[H");
 		fflush(stdout);
 		alternate_buffer = true;
 	}
+#endif
 }
 
 static int client_mainloop(void) {
